### ISSUE TITLE
Stack overflow in clrjit!Compiler::fgMorphSmpOp/clrjit!Compiler::fgMorphTree

### ISSUE THREAD

**Issue by danmosemsft** [dotnet/coreclr#23309]
_From @farlock85 on March 14, 2019 15:53_

Our new application suddenly started throwing StackOverflow exceptions, which were impossible to pinpoint because of the missing stack trace. After spending some time commenting out code I found that there is an issue with dependency injection, specifically when you do a lot of it.

One service had about 500 generic classes injected (think IQueryContext<T> concept), which in turn had 5 other services injected. When doing any API call the class got created, without issues, but when doing a subsequent call it would throw the StackOverflowException. When removing some of the injected services it would throw the StackOverflow exception only after 5-6 API calls, and by removing even more it would stop.

After doing some testing I couldn't reproduce the stack overflow exception but noticed when you inject more than 700 services into a class it will fail in Visual Studio 2017 with an Access Violation. I attached a sample to show this issue.
[TestApp1.zip](https://github.com/dotnet/core/files/2967265/TestApp1.zip)

Now I understand that this is excessive and I re-wrote some of our app to dynamically load the correct service instead, but is there any official stance on the limitations of dependency injection?

_Copied from original issue: dotnet/core#2458_

category:correctness
theme:minopts
skill-level:expert
cost:medium

**Comment by danmosemsft** [comment]
_From @leecow on March 15, 2019 17:41_

@danmosemsft and @jaredpar for possible insights.

**Comment by danmosemsft** [comment]
Codegen.
```
(a50.651c): Stack overflow - code c00000fd (first chance)
First chance exceptions are reported before any exception handling.
This exception may be expected and handled.
clrjit!Compiler::fgMorphSmpOp+0x4d:
00007ffa`f9ba2abd 664489752c      mov     word ptr [rbp+2Ch],r14w ss:00000095`d6043ffc=0000
0:042> k
 # Child-SP          RetAddr           Call Site
00 00000095`d6043f90 00007ffa`f9ba20c3 clrjit!Compiler::fgMorphSmpOp+0x4d [e:\a\_work\307\s\src\jit\morph.cpp @ 11667] 
01 00000095`d6044140 00007ffa`f9ba2d2a clrjit!Compiler::fgMorphTree+0xf3 [e:\a\_work\307\s\src\jit\morph.cpp @ 15933] 
02 00000095`d6044210 00007ffa`f9ba20c3 clrjit!Compiler::fgMorphSmpOp+0x2ba [e:\a\_work\307\s\src\jit\morph.cpp @ 12478] 
03 00000095`d60443c0 00007ffa`f9ba2d2a clrjit!Compiler::fgMorphTree+0xf3 [e:\a\_work\307\s\src\jit\morph.cpp @ 15933] 
04 00000095`d6044490 00007ffa`f9ba20c3 clrjit!Compiler::fgMorphSmpOp+0x2ba [e:\a\_work\307\s\src\jit\morph.cpp @ 12478] 
05 00000095`d6044640 00007ffa`f9ba2d2a clrjit!Compiler::fgMorphTree+0xf3 [e:\a\_work\307\s\src\jit\morph.cpp @ 15933] 

```

**Comment by AndyAyersMS** [comment]
I'll take a look.

**Comment by AndyAyersMS** [comment]
Finally have a repro of the stack overflow -- needed to figure out how to run IIS Express outside of VS.
Not surprisingly we are jitting what is likely a very large dynamically created method... triggering frames in the app are:
```
000000c64523e250 00007fffb9ea31a8 (MethodDesc 00007fffb97cc170 + 0x88 System.Reflection.Emit.DynamicILGenerator.GetCallableMethod(System.Reflection.RuntimeModule, System.Reflection.Emit.DynamicMethod)), calling coreclr!JIT_WriteBarrier [E:\A\_work\307\s\src\vm\amd64\JitHelpers_Fast.asm:466]
000000c64523e260 00007fffb9d95cfa (MethodDesc 00007fffb9703fe8 + 0x7a DomainNeutralILStubClass.IL_STUB_PInvoke(System.IRuntimeMethodInfo))
000000c64523e2b8 00007fffb9d95cfa (MethodDesc 00007fffb9703fe8 + 0x7a DomainNeutralILStubClass.IL_STUB_PInvoke(System.IRuntimeMethodInfo))
000000c64523e310 00007fffb9ea5e27 (MethodDesc 00007fffb97cc5e8 + 0x37 System.Reflection.Emit.DynamicMethod.CreateDelegate(System.Type, System.Object)), calling 00007fffb98e3648 (stub for System.Runtime.CompilerServices.RuntimeHelpers._CompileMethod(System.IRuntimeMethodInfo))
000000c64523e320 00007fffc477ee3f (MethodDesc 00007fff6429c420 + 0x1f System.Linq.Expressions.Expression`1[[System.__Canon, System.Private.CoreLib]].get_TypeCore()), calling coreclr!JIT_GetRuntimeType [e:\a\_work\307\s\src\vm\jithelpers.cpp:4240]
000000c64523e360 00007fffc47c1e70 (MethodDesc 00007fff642c9c28 + 0x70 System.Linq.Expressions.Compiler.LambdaCompiler.CreateDelegate()), calling CLRStub[VSD_DispatchStub]@7fff63c70de0
000000c64523e380 00007fffc47c63db (MethodDesc 00007fff642ca088 + 0x3b System.Linq.Expressions.Compiler.LambdaCompiler.EmitLambdaBody()), calling (MethodDesc 00007fff642ca098 + 0 System.Linq.Expressions.Compiler.LambdaCompiler.EmitLambdaBody(System.Linq.Expressions.Compiler.CompilerScope, Boolean, CompilationFlags))
000000c64523e3b0 00007fffc47c1c5b (MethodDesc 00007fff642c9bc8 + 0x8b System.Linq.Expressions.Compiler.LambdaCompiler.Compile(System.Linq.Expressions.LambdaExpression)), calling (MethodDesc 00007fff642c9c28 + 0 System.Linq.Expressions.Compiler.LambdaCompiler.CreateDelegate())
000000c64523e3f0 00007fffc477ef17 (MethodDesc 00007fff6429c438 + 0x77 System.Linq.Expressions.Expression`1[[System.__Canon, System.Private.CoreLib]].Compile(Boolean)), calling (MethodDesc 00007fff642c9bc8 + 0 System.Linq.Expressions.Compiler.LambdaCompiler.Compile(System.Linq.Expressions.LambdaExpression))
000000c64523e440 00007fffc477ee8d (MethodDesc 00007fff6429c430 + 0xd System.Linq.Expressions.Expression`1[[System.__Canon, System.Private.CoreLib]].Compile()), calling (MethodDesc 00007fff6429c438 + 0 System.Linq.Expressions.Expression`1[[System.__Canon, System.Private.CoreLib]].Compile(Boolean))
000000c64523e470 00007ffff843b20e (MethodDesc 00007fff63f9a5a8 + 0xde Microsoft.Extensions.DependencyInjection.ActivatorUtilities.CreateFactory(System.Type, System.Type[])), calling (MethodDesc 00007fff6429c430 + 0 System.Linq.Expressions.Expression`1[[System.__Canon, System.Private.CoreLib]].Compile())
```
Trying to get a picture of what that method looks like now...

**Comment by AndyAyersMS** [comment]
Looks like the method in question pends a large number of things to the stack and then calls a constructor:
```
IL_0000: ldarg.1 
IL_0001: ldtoken 2000002 "TestApp1.ITest1"
IL_0006: call 6000003 System.Type.GetTypeFromHandle(System.RuntimeTypeHandle)
IL_000b: ldtoken 2000004 "TestApp1.Controllers.ValuesController"
IL_0010: call 6000005 System.Type.GetTypeFromHandle(System.RuntimeTypeHandle)
IL_0015: ldc.i4.0 
IL_0016: call 6000006 Microsoft.Extensions.DependencyInjection.ActivatorUtilities.GetService(System.IServiceProvider, System.Type, System.Type, Boolean)
IL_001b: castclass 2000007 "TestApp1.ITest1"
  ...
IL_6260: ldarg.1 
IL_6261: ldtoken 2001274 "TestApp1.IQueryContext`1[[TestApp1.MyClass784, TestApp1]]"
IL_6266: call 6001275 System.Type.GetTypeFromHandle(System.RuntimeTypeHandle)
IL_626b: ldtoken 2001276 "TestApp1.Controllers.ValuesController"
IL_6270: call 6001277 System.Type.GetTypeFromHandle(System.RuntimeTypeHandle)
IL_6275: ldc.i4.0 
IL_6276: call 6001278 Microsoft.Extensions.DependencyInjection.ActivatorUtilities.GetService(System.IServiceProvider, System.Type, System.Type, Boolean)
IL_627b: castclass 2001279 "TestApp1.IQueryContext`1[[TestApp1.MyClass784, TestApp1]]"
IL_6280: newobj 600127a TestApp1.Controllers.ValuesController..ctor(TestApp1.ITest1, TestApp1.ITest2, TestApp1.ITest3, TestApp1.ITest4, TestApp1.IQueryContext`1<TestApp1.MyClass1>, TestApp1.IQueryContext`1<TestApp1.MyClass2>, TestApp1.IQueryContext`1<TestApp1.MyClass3>, TestApp1.IQueryContext`1<TestApp1.MyClass4>, TestApp1.IQueryContext`1<TestApp1.MyClass5>, TestApp1.IQueryContext`1<TestApp1.MyClass6>, TestApp1.IQueryContext`1<TestApp1.MyClass7>, TestApp1.IQueryContext`1<TestApp1.MyClass8>, TestApp1.IQueryContext`1<TestApp1.MyClass9>, TestApp1.IQueryContext`1<TestApp1.MyClass10>, TestApp1.IQueryContext`1<TestApp1.MyClass11>, TestApp1.IQueryContext`1<TestApp1.MyClass12>, TestApp1.IQueryContext`1<TestApp1.MyClass13>, TestApp1.IQueryContext`1<TestApp1.MyClass14>, TestApp1.IQueryContext`1<TestApp1.MyClass15>, TestApp1.IQueryContext`1<TestApp1.MyClass16>, TestApp1.IQueryContext`1<TestApp1.MyClass17>, TestApp1.IQueryContext`1<TestApp1.MyClass18>, TestApp1.IQueryContext`1<TestApp1.MyClass19>, TestApp1.IQueryContext`1<TestApp1.MyClass20>, TestApp1.IQueryContext`1<TestApp1.MyClass21>, TestApp1.IQueryContext`1<TestApp1.MyClass22>, TestApp1.IQueryContext`1<TestApp1.MyClass23>, TestApp1.IQueryContext`1<TestApp1.MyClass24>, TestApp1.IQueryContext`1<TestApp1.MyClass25>, TestApp1.IQueryContext`1<TestApp1.MyClass26>, TestApp1.IQueryContext`1<TestApp1.MyClass27>, TestApp1.IQueryContext`1<TestApp1.MyClass28>, TestApp1.IQueryContext`1<TestApp1.MyClass29>, TestApp1.IQueryContext`1<TestApp1.MyClass30>, TestApp1.IQueryContext`1<TestApp1.MyClass31>, TestApp1.IQueryContext`1<TestApp1.MyClass32>, TestApp1.IQueryContext`1<TestApp1.MyClass33>, TestApp1.IQueryContext`1<TestApp1.MyClass34>, TestApp1.IQueryContext`1<TestApp1.MyClass35>, TestApp1.IQueryContext`1<TestApp1.MyClass36>, TestApp1.IQueryContext`1<TestApp1.MyClass37>, TestApp1.IQueryContext`1<TestApp1.MyClass38>, TestApp1.IQueryContext`1<TestApp1.MyClass39>, TestApp1.IQueryContext`1<TestApp1.MyClass40>, TestApp1.IQueryContext`1<TestApp1.MyClass41>, TestApp1.IQueryContext`1<TestApp1.MyClass42>, ...
IL_6285: ret 
```
Repro is on 2.2.

Note under IIS stack limit is 512MB.

That .ctor has 785 arguments, seems to parallel the number of services added.

The jit should be breaking up the size of input trees to avoid recursion issues with really big trees. But we avoid splitting trees for certain special calls like `GetTypeFromHandle`. Still, I would expect the spilling to happen when we reach `GetService` or that last ctor call.

Let me work on getting a jit dump.

**Comment by AndyAyersMS** [comment]
Hmm, checked jit does not get the stack overflow.
```
Compiling  504 DynamicClass::lambda_method, IL size = 25222, hsh=0x1fb7cd59
Compiling  505 ResourceInvoker::InvokeAsync, IL size = 57, hsh=0xc97b4a4
```
So wonder if there is some kind of chk-rel divergence?

**Comment by AndyAyersMS** [comment]
Importer-driving splitting seems to be happening, every 0xd0 bytes or so, and it looks to be done the same way in chk and release. So it's not that...

**Comment by AndyAyersMS** [comment]
I suppose even if we split trees the arglist for the constructor is going to have ~700 entries, and because of the clunky way arglists are represented, that would require 1400 nodes. But that recursion should happen the same way in chk and release, and if anything, chk should use more stack.

But, perhaps the ret build is inlining more aggressively, so each frame in RET takes up more space than in CHK...? Seems like something we can measure.

**Comment by AndyAyersMS** [comment]
Note normally we're fairly careful to iterate over the args list, but if we're remorphing, we recursively process them:
```
            call->gtCallLateArgs = fgMorphTree(call->gtCallLateArgs)->AsArgList();
```
And yes frame sizes grow in release builds, especially for `fgMorphSmpOp`:

| Method  |  Check |  Release |
| ------- | ---- | ------ |
| `fgMorphTree` | 160 | 208 |
| `fgMorphSmpOp` | 208 | 432 |

Since these two call each other recursively to morph the late arglist, we get 640 bytes per pair, and 820 of those pairs will blow out 512K of stack (while on checked, we'll only use 287K). And there's already a decent chunk of stuff on the stack when this recursion starts, so that explains why after ~780 or so of these we hit the limit.

**Comment by AndyAyersMS** [comment]
See also the perhaps related #21450 (not clear there what the root cause of overflows is...).

**Comment by AndyAyersMS** [comment]
cc @dotnet/jit-contrib 

Seems like we ought to fix this recursion. More worrying is the fact that we can see overflows in release that we don't see in checked. Not clear what we do about that.

**Comment by mikedn** [comment]
Remains to be seen if fixing this in morph won't just move the stack overflow somewhere else. I'm pretty sure loop hoisting will also recurse down `GT_LIST` nodes (though I suppose this particular call isn't in a loop so perhaps it doesn't matter). There are many reasons I'm trying to get rid of `GT_LIST`...

**Comment by AndyAyersMS** [comment]
Trying to come up with a simpler repro but no luck so far.

If you launch via VS it will not catch the exceptions as it sets itself in managed debug mode. So you need to use windbg, and hence launch from the command line. If you just `dotnet run` it will use a bigger stack. So need to host this in IIS express.

Here's my repro method. Likely fragile. First build the test app via `dotnet build`. Then launch server via:
```
set COREVER=2.2
set IISEXPRESS="c:\Program Files\IIS Express\iisexpress.exe"
set LAUNCHER_ARGS=exec d:\bugs\23309\Orig\TestApp1\bin\Debug\netcoreapp%COREVER%\TestApp1.dll
set LAUNCHER_PATH=dotnet

set ASPNETCORE_ENVIRONMENT=Development
set ASPNETCORE_HOSTINGSTARTUPASSEMBLIES=Microsoft.AspNetCore.Server.IISIntegration
set ASPNETCORE_HTTPS_PORT=44387
set ASPNETCORE_IIS_HTTPAUTH=anonymous;
set ASPNETCORE_IIS_PHYSICAL__PATH=d:\bugs\23309\Orig\TestApp1\

pushd d:\bugs\23309\Orig\TestApp1\

@echo ** Targeting %COREVER% **

%IISEXPRESS% /config:"d:\bugs\23309\Orig\.vs\config\applicationhost.config" /site:"TestApp1" /apppool:"TestApp1 AppPool"
```
For client side, once the server is running:
```
curl  https://localhost:44387/api/values
```
To use a checked jit I have a 2.2 local build and copy the jit from that into the right shared folder.

**Comment by AndyAyersMS** [comment]
Looks like this was fixed in 3.0 via #18582. Perhaps we should think about propagating that to 2.2.

**Comment by AndyAyersMS** [comment]
Porting #18582 to 2.2 fixes the repro case.

**Comment by AndyAyersMS** [comment]
Will wait for #23346 to merge, and then will propose porting that and #18582 over to 2.2.

**Comment by AndyAyersMS** [comment]
@sandreenko the test is merged -- in the 2.2 PR please pick it up as well as the fix.

**Comment by AndyAyersMS** [comment]
@farlock85 thanks for reporting this. We had the issue fixed in master so we'll port that fix back to 2.2.

It is slated to be fixed in 2.2.5.

**Comment by farlock85** [comment]
Appreciate all the work @AndyAyersMS and the rest of the team! I've been following along the conversations.

The only thing I'm not clear on if there is still a limit on dependency injection? I assume there is a technical limit (memory or stacksize?), but are we talking 1000's, 10.000's or even more?

**Comment by AndyAyersMS** [comment]
I think you can probably get close to the implementation limit on the number of method arguments -- something like 16384 or 65536.

**Comment by AndyAyersMS** [comment]
Will keep this open until we land the patch in 2.2.

**Comment by farlock85** [comment]
@AndyAyersMS was this fix release in 2.2.5?

**Comment by AndyAyersMS** [comment]
> was this fix release in 2.2.5?

Yes, it should be fixed in 2.2.5.

When you've had a chance to try it, let me know one way or the other.

**Comment by farlock85** [comment]
It seems to have fixed my test case application. Unfortunately when I restore our application back to its original DI state it still fails with a stack overflow exception. I have been trying to figure out a way to estimate how many injections are actually happening but without being able to access an actual counter it seems near impossible. My guess is that's easily going over 16k injectables.

 A large part of the application was rewritten to do dynamic instantiating and that seems to work fine for us.

**Comment by AndyAyersMS** [comment]
Can you get a stack trace at the point of overflow, or share a dump file? Could be we need similar fixes in other places, and for some reason the test case in this issue doesn't run into them all.

**Comment by farlock85** [comment]
I'm not able to get a stack trace, this is the call stack:
```
[Lightweight Function]
Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(System.Type serviceType, Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope serviceProviderEngineScope)
Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(System.Type serviceType)
Microsoft.Extensions.DependencyInjection.Abstractions.dll!Microsoft.Extensions.DependencyInjection.ActivatorUtilities.GetService(System.IServiceProvider sp, System.Type type = {Name = Inspecting the state of an object in the debuggee of type System.Type is not supported in this context. FullName = Inspecting the state of an object in the debuggee of type System.Type is not supported in this context.}, System.Type requiredBy = {Name = Inspecting the state of an object in the debuggee of type System.Type is not supported in this context. FullName = Inspecting the state of an object in the debuggee of type System.Type is not supported in this context.}, bool isDefaultParameterRequired = false)
[Lightweight Function]
Microsoft.AspNetCore.Mvc.Core.dll!Microsoft.AspNetCore.Mvc.Controllers.ControllerActivatorProvider.CreateActivator.AnonymousMethod__0(Microsoft.AspNetCore.Mvc.ControllerContext controllerContext)
Microsoft.AspNetCore.Mvc.Core.dll!Microsoft.AspNetCore.Mvc.Controllers.ControllerFactoryProvider.CreateControllerFactory.__CreateController|0(Microsoft.AspNetCore.Mvc.ControllerContext controllerContext = {Microsoft.AspNetCore.Mvc.ControllerContext})
Microsoft.AspNetCore.Mvc.Core.dll!Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Next(ref Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.State next = ActionBegin, ref Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Scope scope = Invoker, ref object state = null, ref bool isCompleted)
Microsoft.AspNetCore.Mvc.Core.dll!Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeInnerFilterAsync()
System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start<Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeInnerFilterAsync>d__13>(ref Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeInnerFilterAsync>d__13 stateMachine)
Microsoft.AspNetCore.Mvc.Core.dll!Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeInnerFilterAsync()
Microsoft.AspNetCore.Mvc.Core.dll!Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(ref Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.State next = ResourceNext, ref Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Scope scope = Resource, ref object state, ref bool isCompleted = false)
Microsoft.AspNetCore.Mvc.Core.dll!Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResourceFilter()
System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start<Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.<InvokeNextResourceFilter>d__23>(ref Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.<InvokeNextResourceFilter>d__23 stateMachine)
Microsoft.AspNetCore.Mvc.Core.dll!Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResourceFilter()
Microsoft.AspNetCore.Mvc.Core.dll!Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(ref Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.State next = InvokeBegin, ref Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Scope scope = Invoker, ref object state = {Microsoft.AspNetCore.Mvc.ViewFeatures.Internal.SaveTempDataFilter}, ref bool isCompleted = false)
Microsoft.AspNetCore.Mvc.Core.dll!Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeFilterPipelineAsync()
System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start<Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.<InvokeFilterPipelineAsync>d__18>(ref Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.<InvokeFilterPipelineAsync>d__18 stateMachine)
Microsoft.AspNetCore.Mvc.Core.dll!Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeFilterPipelineAsync()
Microsoft.AspNetCore.Mvc.Core.dll!Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeAsync()
System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start<Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.<InvokeAsync>d__16>(ref Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.<InvokeAsync>d__16 stateMachine)
Microsoft.AspNetCore.Mvc.Core.dll!Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeAsync()
Microsoft.AspNetCore.Mvc.Core.dll!Microsoft.AspNetCore.Mvc.Internal.MvcEndpointDataSource.CreateEndpoint.AnonymousMethod__0(Microsoft.AspNetCore.Http.HttpContext context)
Microsoft.AspNetCore.Routing.dll!Microsoft.AspNetCore.Routing.EndpointMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext httpContext = {Microsoft.AspNetCore.Http.DefaultHttpContext})
System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start<Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>d__3>(ref Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>d__3 stateMachine)
Microsoft.AspNetCore.Routing.dll!Microsoft.AspNetCore.Routing.EndpointMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext httpContext)
Microsoft.AspNetCore.Routing.dll!Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext httpContext = {Microsoft.AspNetCore.Http.DefaultHttpContext})
System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start<Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware.<Invoke>d__6>(ref Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware.<Invoke>d__6 stateMachine)
Microsoft.AspNetCore.Routing.dll!Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext httpContext)
Microsoft.AspNetCore.Cors.dll!Microsoft.AspNetCore.Cors.Infrastructure.CorsMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext context)
Microsoft.AspNetCore.StaticFiles.dll!Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext context = {Microsoft.AspNetCore.Http.DefaultHttpContext})
System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start<Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.<Invoke>d__7>(ref Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.<Invoke>d__7 stateMachine)
Microsoft.AspNetCore.StaticFiles.dll!Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext context)
Microsoft.AspNetCore.StaticFiles.dll!Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext context = {Microsoft.AspNetCore.Http.DefaultHttpContext})
System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start<Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.<Invoke>d__7>(ref Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.<Invoke>d__7 stateMachine)
Microsoft.AspNetCore.StaticFiles.dll!Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext context)
Microsoft.AspNetCore.StaticFiles.dll!Microsoft.AspNetCore.StaticFiles.DefaultFilesMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext context)
Microsoft.AspNetCore.Authentication.dll!Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext context = {Microsoft.AspNetCore.Http.DefaultHttpContext})
System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start<Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.<Invoke>d__6>(ref Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.<Invoke>d__6 stateMachine)
Microsoft.AspNetCore.Authentication.dll!Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext context)
Microsoft.AspNetCore.Session.dll!Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext context = {Microsoft.AspNetCore.Http.DefaultHttpContext})
System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start<Microsoft.AspNetCore.Session.SessionMiddleware.<Invoke>d__9>(ref Microsoft.AspNetCore.Session.SessionMiddleware.<Invoke>d__9 stateMachine)
Microsoft.AspNetCore.Session.dll!Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext context)
SoapCore.dll!SoapCore.SoapEndpointMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext httpContext = {Microsoft.AspNetCore.Http.DefaultHttpContext}, System.IServiceProvider serviceProvider = {Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope})
System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start<SoapCore.SoapEndpointMiddleware.<Invoke>d__12>(ref SoapCore.SoapEndpointMiddleware.<Invoke>d__12 stateMachine)
SoapCore.dll!SoapCore.SoapEndpointMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext httpContext, System.IServiceProvider serviceProvider)
Microsoft.AspNetCore.Http.Abstractions.dll!Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.UseMiddleware.AnonymousMethod__2(Microsoft.AspNetCore.Http.HttpContext context)
Microsoft.AspNetCore.Diagnostics.dll!Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext context = {Microsoft.AspNetCore.Http.DefaultHttpContext})
System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start<Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.<Invoke>d__7>(ref Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.<Invoke>d__7 stateMachine)
Microsoft.AspNetCore.Diagnostics.dll!Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext context)
Microsoft.AspNetCore.Server.IISIntegration.dll!Microsoft.AspNetCore.Server.IISIntegration.IISMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext httpContext = {Microsoft.AspNetCore.Http.DefaultHttpContext})
System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start<Microsoft.AspNetCore.Server.IISIntegration.IISMiddleware.<Invoke>d__13>(ref Microsoft.AspNetCore.Server.IISIntegration.IISMiddleware.<Invoke>d__13 stateMachine)
Microsoft.AspNetCore.Server.IISIntegration.dll!Microsoft.AspNetCore.Server.IISIntegration.IISMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext httpContext)
Microsoft.AspNetCore.HttpOverrides.dll!Microsoft.AspNetCore.HttpOverrides.ForwardedHeadersMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext context)
Microsoft.AspNetCore.HostFiltering.dll!Microsoft.AspNetCore.HostFiltering.HostFilteringMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext context)
Microsoft.AspNetCore.Hosting.dll!Microsoft.AspNetCore.Hosting.Internal.RequestServicesContainerMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext httpContext)
Microsoft.AspNetCore.Hosting.dll!Microsoft.AspNetCore.Hosting.Internal.HostingApplication.ProcessRequestAsync(Microsoft.AspNetCore.Hosting.Internal.HostingApplication.Context context)
Microsoft.AspNetCore.Server.Kestrel.Core.dll!Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests<Microsoft.AspNetCore.Hosting.Internal.HostingApplication.Context>(Microsoft.AspNetCore.Hosting.Server.IHttpApplication<Microsoft.AspNetCore.Hosting.Internal.HostingApplication.Context> application = {Microsoft.AspNetCore.Hosting.Internal.HostingApplication})
System.Private.CoreLib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state)
System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.Threading.Tasks.VoidTaskResult>.AsyncStateMachineBox<Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.<ProcessRequests>d__185<Microsoft.AspNetCore.Hosting.Internal.HostingApplication.Context>>.MoveNext()
System.Private.CoreLib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state)
System.Private.CoreLib.dll!System.Threading.ThreadPoolWorkQueue.Dispatch()
```

The dump is rather large, could I send you a wetransfer with the file?

**Comment by janvorli** [comment]
> The dump is rather large

@farlock85 gists are good for large logs / dumps. There is a "create gist" in the drop down menu located at the large "plus" character next to your image on the top right corner of this page.

**Comment by farlock85** [comment]
@janvorli the dump is 500MB, I don't see an option in a gist to upload a raw file.

**Comment by janvorli** [comment]
@farlock85 ah, I am sorry for misunderstanding. I have thought you were talking about a textual dump.

**Comment by AndyAyersMS** [comment]
@leecow can you suggest a good way for @farlock85 to share a large dump file?

**Comment by leecow** [comment]
The best bet is probably to open a [VS Dev Community](https://developercommunity.visualstudio.com/content/problem/post.html?space=61) issue. It includes a file attachment option which should do the trick.

**Comment by farlock85** [comment]
Probably easier just to use our OneDrive :)
https://corecreative-my.sharepoint.com/:u:/g/personal/roger_versluis_fieldcap_ca/EaeRpn94-U5AudwtDJTOPogBfLkp8VGohAIIint7imu2aw?e=xBFxOd

**Comment by AndyAyersMS** [comment]
This is now hitting an overflow in the code the jit generates -- specifically, in the prolog stack tickling  (which tries to ensure all needed stack pages for a method are touched before the method runs):
```
DynamicClass.lambda_method(System.Runtime.CompilerServices.Closure, Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope):
00007ffc`1d290080 55              push    rbp
00007ffc`1d290081 57              push    rdi
00007ffc`1d290082 56              push    rsi
00007ffc`1d290083 33c0            xor     eax,eax
00007ffc`1d290085 48850404        test    qword ptr [rsp+rax],rax
00007ffc`1d290089 482d00100000    sub     rax,1000h
00007ffc`1d29008f 483d600ae7ff    cmp     rax,0FFFFFFFFFFE70A60h
```
That big hex constant is `-1,635,744‬` so this method needs about 1.5 MB of stack to execute.

I will look back at the simple example to get some idea why the jit is creating such a large frame, but I think I already know: when methods become very large the jit disables optimization, and this creates lots of stack temporaries. Perhaps there is some simple way to address this -- sometimes we can re-use temporaries to save space.

This stack usage is related to the number of injectables, so the only workaround is to have less of them. Sounds like you already have accomplished this via a rewrite.

**Comment by BruceForstall** [comment]
@AndyAyersMS It sounds like this issue now boils down to "Use less stack space in MinOpts by creating fewer temps". Is this issue worth keeping open? Would it be better to create a new issue to track this?

**Comment by AndyAyersMS** [comment]
There's already an issue for this (use fewer temps when not optimizing) somewhere. I'll try and track it down when I get a chance.

**Comment by AndyAyersMS** [comment]
There are a number of possible causes.

#14890 is the likely culprit here, so we could probably close this as a dup of that.

**Comment by BruceForstall** [comment]
Closing as dup of #14890.
